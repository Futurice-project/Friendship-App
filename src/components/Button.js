import React from 'react';
import styled from 'styled-components/native';

/* Styling for button wrapper */
const ButtonWrapper = styled.TouchableOpacity`
  ${props => {
    if (props.primary) {
      return (
        'background-color: rgba(' +
        (props.color === 'light' ? '255, 255, 255, 1' : '94, 104, 109, 1') +
        ');'
      );
    }
  }};

  ${props => {
    if (props.border) {
      return (
        'border : 2px solid ' +
        (props.color === 'light' ? '#ffffff' : '#60686d') +
        ';'
      );
    }
  }};

  height: 55px;
  width: ${props => {
    return props.size === 'half' ? '160px' : '270px';
  }};
  border-radius: 50px;
`;

/* Styling for text wrapper */
const TextWrapper = styled.View`
  padding: 16px;
  align-items: center;
  justify-content: center;
`;

/* Styling for button text */
const ButtonText = styled.Text`
  font-family: 'NunitoSans-SemiBold'
  color: ${props => {
    switch (props.textColor) {
      case 'white':
        return '#ffffff';
      case 'black':
        return '#60686d';
      default:
        return '#87df91';
    }
  }};
  ${props => {
    if (props.underlined) {
      return 'text-decoration-line: underline;';
    }
  }};
  background-color: transparent;
`;

/**
 * Styled button component
 * @param {boolean} primary - Indicating button nature
 * @param {boolean} border - Enable or disable a border
 * @param {string} title - Button's text
 * @param {string} color - Either "light" or "dark"
 * @param {boolean} underlined - Indicating if the button's text is underlined
 * @param {string} size - Either "half" or "full"
 * @param {string} textColor - Either "white", "black" or "green"
 */
export default class Button extends React.Component {
  render = () => (
    <ButtonWrapper
      primary={this.props.primary}
      color={this.props.color}
      border={this.props.border}
      size={this.props.size}
      onPress={this.props.onPress}
    >
      <TextWrapper>
        <ButtonText
          underlined={this.props.underlined}
          textColor={this.props.textColor}
        >
          {this.props.title}
        </ButtonText>
      </TextWrapper>
    </ButtonWrapper>
  );
}
