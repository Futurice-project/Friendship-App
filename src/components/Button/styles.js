import { StyleSheet } from 'react-native';
import { colors, fonts, fontSizes, paddings } from '../../styles';

export default StyleSheet.create({
  button: {
    borderRadius: 50,
    alignItems: 'center',
    justifyContent: 'center',
    paddingVertical: paddings.XS,
  },
  buttonSecondary: {
    borderRadius: 50,
    alignItems: 'center',
    justifyContent: 'center',
    paddingVertical: paddings.XS,
    borderWidth: 2,
    borderColor: colors.BEIGE,
  },
  headerButton: {
    backgroundColor: 'transparent',
    alignItems: 'center',
    justifyContent: 'center',
  },
  floatingButton: {
    borderRadius: 50,
    alignItems: 'center',
    justifyContent: 'center',
    paddingHorizontal: paddings.XXS,
  },
  text: {
    color: colors.BEIGE,
    fontSize: fontSizes.BODY_TEXT,
    fontFamily: fonts.REGULAR,
  },
  buttonText: {
    color: colors.ORANGE,
    fontSize: fontSizes.BODY_TEXT,
    fontFamily: fonts.REGULAR,
  },
});
