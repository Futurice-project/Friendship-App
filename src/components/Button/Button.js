import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Text, TouchableOpacity } from 'react-native';
import styles from './styles';
import { colors } from '../../styles';

class Button extends Component {
  render() {
    const { primary, text, width, onPress, header, color } = this.props;

    let style = primary
      ? styles.button
      : header ? styles.headerButton : styles.buttonSecondary;
    let buttonTextColor;

    switch (color) {
      case 'orange':
        buttonTextColor = colors.ORANGE;
        break;
      case 'white':
        buttonTextColor = colors.WHITE;
        break;
    }

    let buttonWidth;

    switch (width) {
      case 'sm':
        buttonWidth = 100;
        break;
      case 'md':
        buttonWidth = 150;
        break;
      case 'xl':
        buttonWidth = 200;
        break;
      case 'noFixedWidth':
        break;
    }

    return (
      <TouchableOpacity
        style={[style, { width: buttonWidth }]}
        onPress={onPress}
      >
        {this.props.icon ? (
          this.props.icon
        ) : (
          <Text style={[styles.text, { color: buttonTextColor }]}>{text}</Text>
        )}
      </TouchableOpacity>
    );
  }
}

Button.propTypes = {
  width: PropTypes.string,
  text: PropTypes.string,
  primary: PropTypes.bool,
  color: PropTypes.string,
};

Button.defaultProps = {
  text: 'A Button',
  primary: false,
  width: 'noFixedWidth',
  color: 'orange',
  onPress: () => {},
};

export default Button;
