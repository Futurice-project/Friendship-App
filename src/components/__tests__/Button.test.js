import React from 'react';
import { shallow } from 'enzyme';

import { ButtonWrapper } from '../Button';

describe('ButtonWrapper', () => {
  test('renders correctly', () => {
    const wrapper = shallow(<ButtonWrapper />);

    expect(wrapper).toMatchSnapshot();
  });

  test('props change the design correctly', () => {
    // we use shallow rendering to test our components
    // http://airbnb.io/enzyme/docs/api/shallow.html
    const button = shallow(<ButtonWrapper />);
    const buttonBorder = shallow(<ButtonWrapper border />);
    const buttonPrimary = shallow(<ButtonWrapper primary />);
    const buttonPrimaryLight = shallow(<ButtonWrapper primary color="light" />);

    // see: http://airbnb.io/enzyme/docs/api/ShallowWrapper/debug.html
    // console.log(button.debug());
    // console.log(buttonPrimary.debug());

    // you can see in the console that the backgroundColor is different for each of them
    // console.log('button props:', button.props());
    // console.log('buttonPrimary props:', buttonPrimary.props());
    // console.log('buttonPrimaryLight props', buttonPrimaryLight.props());
    // console.log('buttonBorder props', buttonBorder.props());

    // console.log('button state: ', button.state());
    // console.log('buttonPrimary state:', buttonPrimary.state());
    // console.log('buttonPrimaryLight state:', buttonPrimaryLight.state());

    // console.log(button.props().style[0].backgroundColor);
    // console.log(buttonPrimary.props().style[0].backgroundColor);
    // console.log(buttonPrimaryLight.props().style[0].backgroundColor);

    expect(button.props().style[0].backgroundColor).toBe(undefined);
    expect(buttonPrimary.props().style[0].backgroundColor).toBe(
      'rgba(94, 104, 109, 1)',
    );
    expect(buttonPrimaryLight.props().style[0].backgroundColor).toBe(
      'rgba(255, 255, 255, 1)',
    );

    expect(buttonBorder.props().style[0].borderWidth).toBe(2);
  });
});
