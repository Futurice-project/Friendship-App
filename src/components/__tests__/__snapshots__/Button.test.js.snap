// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonWrapper renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Styled(TouchableOpacity) />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "activeOpacity": 0.2,
      "children": undefined,
      "style": Array [
        Object {
          "borderRadius": 50,
          "height": 55,
          "width": 270,
        },
        undefined,
      ],
    },
    "ref": [Function],
    "rendered": null,
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "activeOpacity": 0.2,
        "children": undefined,
        "style": Array [
          Object {
            "borderRadius": 50,
            "height": 55,
            "width": 270,
          },
          undefined,
        ],
      },
      "ref": [Function],
      "rendered": null,
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
